packages/block-editor/src/components/block-list/block.js

```diff
@@ -49,108 +49,181 @@ import Inserter from '../inserter';
 import HoverArea from './hover-area';
 import { isInsideRootBlock } from '../../utils/dom';
 
-export class BlockListBlock extends Component {
-	constructor() {
-		super( ...arguments );
-
-		this.setBlockListRef = this.setBlockListRef.bind( this );
-		this.bindBlockNode = this.bindBlockNode.bind( this );
-		this.setAttributes = this.setAttributes.bind( this );
-		this.maybeHover = this.maybeHover.bind( this );
-		this.forceFocusedContextualToolbar = this.forceFocusedContextualToolbar.bind( this );
-		this.hideHoverEffects = this.hideHoverEffects.bind( this );
-		this.onFocus = this.onFocus.bind( this );
-		this.preventDrag = this.preventDrag.bind( this );
-		this.onPointerDown = this.onPointerDown.bind( this );
-		this.deleteOrInsertAfterWrapper = this.deleteOrInsertAfterWrapper.bind( this );
-		this.onBlockError = this.onBlockError.bind( this );
-		this.onTouchStart = this.onTouchStart.bind( this );
-		this.onClick = this.onClick.bind( this );
-		this.onDragStart = this.onDragStart.bind( this );
-		this.onDragEnd = this.onDragEnd.bind( this );
-		this.selectOnOpen = this.selectOnOpen.bind( this );
-		this.hadTouchStart = false;
-
-		this.state = {
-			error: null,
-			dragging: false,
-			isHovered: false,
-		};
-		this.isForcingContextualToolbar = false;
-	}
+/**
+ * Prevents default dragging behavior within a block to allow for multi-
+ * selection to take effect unhampered.
+ *
+ * @param {DragEvent} event Drag event.
+ *
+ * @return {void}
+ */
+const preventDrag = ( event ) => {
+	event.preventDefault();
+};
+
+function BlockListBlock( {
+	blockRef,
+	mode,
+	isFocusMode,
+	hasFixedToolbar,
+	isLocked,
+	isFirst,
+	isLast,
+	clientId,
+	rootClientId,
+	isSelected,
+	isPartOfMultiSelection,
+	isFirstMultiSelected,
+	isTypingWithinBlock,
+	isCaretWithinFormattedText,
+	isEmptyDefaultBlock,
+	isMovable,
+	isParentOfSelectedBlock,
+	isDraggable,
+	isSelectionEnabled,
+	className,
+	name,
+	isValid,
+	attributes,
+	initialPosition,
+	wrapperProps,
+	setAttributes,
+	onReplace,
+	onInsertBlocksAfter,
+	onMerge,
+	onSelect,
+	onRemove,
+	onInsertDefaultBlockAfter,
+	toggleSelection,
+	onShiftSelection,
+	onSelectionStart,
+} ) {
+	// Random state used to rerender the component if needed, ideally we don't need this
+	const [ , updateRerenderState ] = useState( {} );
+	const rerender = () => updateRerenderState( {} );
+
+	// Reference of the wrapper -------------------------------------------------
+	const wrapper = useRef( null );
+	useEffect( () => {
+		blockRef( wrapper.current, clientId );
+		// We need to rerender to trigger a rerendering of HoverArea.
+		rerender();
+	}, [] );
```

Comment:
I see ESLint warnings here about needing to define `wrapper` and `clientId` as dependencies.

Aside: Is there a reason these ESLint rules are defined as warnings, not errors?

Reply:
After some hooks explorations in this PR and others, I think I'm going to remove this rule entirely as there are valid use-cases to omit dependencies (I want to only run this effect once). I feel that's why the React team suggests to use warnings here.

Reply:
I'm not sure your should remove the rule entirely but maybe just ignore the rule for the valid uses?  Hooks are still fairly new to the ecosystem and warnings can help contributors spot improper uses.

Reply:
I did three refactorings to use hooks, in all of them, I found myself violating this rule. We can reconsider if there's an alternative way of using hooks, something I didn't discover yet for these use-cases which IMO are common.

Reply:
I'm still wary about usage of this pattern.  React Docs warn pretty strongly against it and Dan Abramov uses pretty strict warnings about it in an [article he dedicated to `useEffect`](https://overreacted.io/a-complete-guide-to-useeffect/):

> “But I only want to run it on mount!”, you’ll say. For now, remember: if you specify deps, all values from inside your component that are used by the effect must be there. Including props, state, functions — anything in your component.

Could it be written something like this?

```js
const wrapper = useRef( null );
const callBlockRef = () => blockRef( wrapper.current, clientId );
useEffect( () => {
    callBlockRef();
    // we need to rerender to trigger a rerendering of Hover Area.
    rerender();
} );
```

Reply:
Sorry my example is bad, something like this maybe:

```js
const callBlockRef = useCallback( ( wrapper, client.Id ) => {
    blockRef( wrapper.current, clientId );
}, [] );
useEffect( () => {
   callBlockRef();
  // we need to rerender to trigger a rerendering of Hover Area.
  rerender();
}, [ callBlockRef ] );
```

Reply:
I agree this is not great code, but I think the fix for this is to not use `blockRef` prop entirely and just use ref forwarding to get the node from the block list. I tried doing this and I hit a wall where for some reason we're not ready yet to do it (I don't recall the exact reason as it was two weeks ago, probably related to BlockList not being hook ready)